File: Data Structures & Algorithms/reorder-linked-list/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        temp = []
        curr = head

        while curr:
            temp.append(curr)
            curr = curr.next

        left, right = 0, len(temp) - 1

        while left < right:
            temp[left].next = temp[right]
            left += 1
            temp[right].next = temp[left]
            right -= 1

        temp[left].next = None

File: Data Structures & Algorithms/linked-list-cycle-detection/submission-1.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    bool hasCycle(ListNode* head) {
        std::unordered_set<ListNode*> set;  
        ListNode* ptr = head;

        while (ptr) {
           if (set.find(ptr) != set.end()) {
            return true;  
           } else {
            set.insert(ptr); 
            ptr = ptr->next; 
           } 
        } 
        return false; 

    }
};
